[#10938] Support copying courses

ypinhsuan opened the pull request:
Part of #10938, Closes #11012

**Outline of Solution**
Original credits to @LeeBadal, @heeenkie, @Grunler, @MatildaAR, @Olle251 for implementation and frontend tests

A copy course option similar to copy feedback sessions
Added an option to allow instructors to copy from other courses when adding new course

There should be a dropdown list for `Course to copy from` but somehow my screen recording is not capturing it
![gif](https://user-images.githubusercontent.com/58677983/120891071-24eed700-c639-11eb-800b-639d0a86cdd7.gif)
![TEAMMATES-Online-Peer-Feedback-E](https://user-images.githubusercontent.com/58677983/120891069-228c7d00-c639-11eb-9ac1-7ef25172fc89.gif)

daongochieu2810 commented on src/web/app/components/copy-course-modal/copy-course-modal.component.ts:
```diff
@@ -0,0 +1,110 @@
+   * Clears all selected feedback sessions.
+   */
+  clearSelectedFeedbackSession(): void {
+    this.chosenFeedbackSessions = new Set<FeedbackSession>();
```
you can use `this.chosenFeedbackSessions.clear()` here

daongochieu2810 commented on src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:
```diff
@@ -100,4 +108,30 @@ export class AddCourseFormComponent implements OnInit {
+  /**
+   * Handles copying from other courses event
+   */
+  onCopy(): void {
```
I think we should declare this method as `async` so it will not block for long in case of coying multiple big courses

moziliar replied:
This at the same time means we are also allowing concurrent connection to the backend. How's the resource consumption of this change?

daongochieu2810 replied:
yes I agree with disabling the copy course button while copying. Or we can turn it into a cancel button

moziliar replied:
This sounds like a UX question😅I'm no expert in it, but just raising the resource issue.
Can do either way! Cancel button might involve rollback btw.

ypinhsuan replied:
Is this ok? The copy button is disabled when copying.
![ezgif com-gif-maker](https://user-images.githubusercontent.com/58677983/120914086-48b02c80-c6ce-11eb-85eb-82497e7757ef.gif)

moziliar replied:
Sorry for not getting back on this.
@damithc's take on loading UI would be to show the user some progress when it takes sometime to load.
Is there a way to perform this? I remember we are copying session by session (in the future we might need to improve this, but so far no such need as this is not a frequent use case).

ypinhsuan replied:
Added a progress bar to show users the progress
![New-video](https://user-images.githubusercontent.com/58677983/121646106-8b577780-cac7-11eb-9efa-d05a57bebc5e.gif)

moziliar replied:
This is awesome! I love it!

@damithc prof, are you satisfied with this?
It might not stay should this feature becomes too popular and we subsequently make the endpoint granular (copy course and session all in one API call). But before that, I'd want to see this myself actually.

daongochieu2810 commented on src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:
```diff
@@ -100,4 +108,30 @@ export class AddCourseFormComponent implements OnInit {
+          this.statusMessageService.showErrorToast(resp.error.message);
+        });
+
+        modalRef.result.then((result: CopyCourseModalResult) => this.copyCourseEvent.emit(result), () => {});
```
use `await` after declaring this method as `async`

daongochieu2810 reviewed (commented):
Very good overall! Left some comments

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -246,6 +261,138 @@ export class InstructorCoursesPageComponent implements OnInit {
+  private copyFeedbackSession(fromFeedbackSession: FeedbackSession, newSessionName: string, newCourseId: string):
+      Observable<FeedbackSession> {
+    let createdFeedbackSession!: FeedbackSession;
+    return this.feedbackSessionsService.createFeedbackSession(newCourseId, {
+      feedbackSessionName: newSessionName,
+      instructions: fromFeedbackSession.instructions,
+
+      submissionStartTimestamp: fromFeedbackSession.submissionStartTimestamp,
+      submissionEndTimestamp: fromFeedbackSession.submissionEndTimestamp,
+      gracePeriod: fromFeedbackSession.gracePeriod,
+
+      sessionVisibleSetting: fromFeedbackSession.sessionVisibleSetting,
+      customSessionVisibleTimestamp: fromFeedbackSession.customSessionVisibleTimestamp,
+
+      responseVisibleSetting: fromFeedbackSession.responseVisibleSetting,
+      customResponseVisibleTimestamp: fromFeedbackSession.customResponseVisibleTimestamp,
+
+      isClosingEmailEnabled: fromFeedbackSession.isClosingEmailEnabled,
+      isPublishedEmailEnabled: fromFeedbackSession.isPublishedEmailEnabled,
```
Discussion: is there a way to replace this with something like `feedbackSession.toCreationReqWithName(newSessionName)`?

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -246,6 +261,138 @@ export class InstructorCoursesPageComponent implements OnInit {
+        return from(response.questions).pipe(
+          concatMap((feedbackQuestion: FeedbackQuestion) => {
+            return this.feedbackQuestionsService.createFeedbackQuestion(
+              createdFeedbackSession.courseId, createdFeedbackSession.feedbackSessionName, {
+                questionNumber: feedbackQuestion.questionNumber,
+                questionBrief: feedbackQuestion.questionBrief,
+                questionDescription: feedbackQuestion.questionDescription,
+
+                questionDetails: feedbackQuestion.questionDetails,
+                questionType: feedbackQuestion.questionType,
+
+                giverType: feedbackQuestion.giverType,
+                recipientType: feedbackQuestion.recipientType,
+
+                numberOfEntitiesToGiveFeedbackToSetting: feedbackQuestion.numberOfEntitiesToGiveFeedbackToSetting,
+                customNumberOfEntitiesToGiveFeedbackTo: feedbackQuestion.customNumberOfEntitiesToGiveFeedbackTo,
+
+                showResponsesTo: feedbackQuestion.showResponsesTo,
+                showGiverNameTo: feedbackQuestion.showGiverNameTo,
+                showRecipientNameTo: feedbackQuestion.showRecipientNameTo,
+              });
```
Likewise, any possibility to abstract away this?
Otherwise, it is fine to leave as is; just somewhat unnecessarily chunky here reading as "copying everything from `feedbackQuestion` to a new struct that inherits all the field values => semantically equivalent to a `clone()`

ypinhsuan replied:
ok! will split this into a separate function too

moziliar reviewed (changes requested):
Thanks for the original authors' efforts and @ypinhsuan's effort in curating it towards completion.

Really nicely written code from what I read. Would've been even nicer to have a gif/video to showcase the entire workflow in this PR too.

Have some comments below and some points for discussion.

ypinhsuan commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -246,6 +261,138 @@ export class InstructorCoursesPageComponent implements OnInit {
+   * Creates a copy of a course including all sessions.
+   */
+  onCopy(courseId: string, courseName: string, timeZone: string): void {
+    if (!courseId) {
+      this.statusMessageService.showErrorToast(`Course ${courseId} is not found!`);
+      return;
+    }
```
I don't think there will be such a case where `courseId` is a falsy value, so I will remove this part

ypinhsuan commented:
Added gif in PR description and some frontend tests

moziliar commented on src/web/app/components/copy-course-modal/copy-course-modal.component.ts:
```diff
@@ -105,6 +107,8 @@ export class CopyCourseModalComponent implements OnInit {
+    this.chosenFeedbackSessions.clear();
   }
+
+  private zeroPad: any = (num: number) => String(num).padStart(2, '0');
```
Am wondering if this can be outside the class? It is also fine being inside too:p

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:
```diff
@@ -131,7 +132,9 @@ export class AddCourseFormComponent implements OnInit {
         });
 
-        modalRef.result.then((result: CopyCourseModalResult) => this.copyCourseEvent.emit(result), () => {});
+        await modalRef.result
```
Hmm can I clarify which function call inside is async? `showErrorToast`?

ypinhsuan replied:
I am not very sure but my understanding of `await` is that it is used on `Promise` objects and in this case `modalRef.result` is a Promise. 
```
        try {
          await modalRef.result
          const result: CopyCourseModalResult = await modalRef.result;
          this.copyCourseEvent.emit(result)
        } catch (resp: any) {
          this.statusMessageService.showErrorToast(resp.error.message);
        }
```
Would it be correct to change this part to the above?

moziliar replied:
As in: is there any reason to make this a blocking call?
`subscribe` itself is already a Promise-ish thing. I don't see why you need an await here even tho `modalRef.result` is a Promise.

ypinhsuan replied:
Ok I see, then I think the async and await is not needed here.

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -262,14 +263,9 @@ export class InstructorCoursesPageComponent implements OnInit {
-      this.statusMessageService.showErrorToast(`Course ${courseId} is not found!`);
-      return;
-    }
-
```
Sorry regarding this: do you think there should still be a defensive mechanism in place?
Was commenting on this as the interpolated `courseId` is most likely redundant, but I wasn't rejecting the possibility of such error happening.
Can think of other places that might reuse this function.

ypinhsuan replied:
I see. I agree it will be better to have defensive mechanism

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -350,47 +385,31 @@ export class InstructorCoursesPageComponent implements OnInit {
+  /**
+   * Creates a FeedbackQuestionCreateRequest by cloning the given feedback question.
+   */
+  private fbQuestionCreationReq(feedbackQuestion: FeedbackQuestion): FeedbackQuestionCreateRequest {
```
`toFbQuestionCreationReq` or `newFbQuestionCreationReq`? Making it an action sounding name is probably better.

moziliar reviewed (dismissed):
Added some more comments. Can discuss against my comments if they don't make sense:)
I am largely ok with the current logic. Might need to do some user testing as well before finalising it.

Do resolve the review threads too.

On another note: it is probably better to have more meaningful comment names than `Apply suggested changes`; for the reviewer, I can review by commit to see the exact changes under the commit name; for a developer, I can also `reset` or `revert` certain changes easily with a clearer defined bound.
I know it is hard to keep this up all the time, and that this PR is likely gonna be squashed:p but it's a good habit to have, and that the PR history is still around after squashed merge and can let future maintainer to see exactly what happened.

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/add-course-form/add-course-form.component.ts:
```diff
@@ -100,4 +109,35 @@ export class AddCourseFormComponent implements OnInit {
+        modalRef.componentInstance.isCopyFromOtherSession = true;
+        modalRef.componentInstance.courses = courses.courses;
+
+        courses.courses.forEach((course: Course) => {
+          this.feedbackSessionsService
+            .getFeedbackSessionsForInstructor(course.courseId)
+            .subscribe((feedbackSessions: FeedbackSessions) => {
+              modalRef.componentInstance.courseToFeedbackSession[course.courseId]
+                = [...feedbackSessions.feedbackSessions];
+            });
+        }, (resp: ErrorMessageOutput) => {
+          this.statusMessageService.showErrorToast(resp.error.message);
+        });
```
Looks like it will fan out into many API call at once. I'd keep an eye on it.
What if there are too many sessions to copy? Will it cause new instances to be spun up?
@rrtheonlyone

moziliar commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -246,6 +266,172 @@ export class InstructorCoursesPageComponent implements OnInit {
+          .subscribe((course: Course) => {
+            this.activeCourses.push(this.getCourseModelFromCourse(course));
+            this.activeCoursesDefaultSort();
+            this.isCopyingCourse = false;
```
Btw, when you are copying feedback sessions, is it also considered `CopyingCourse`?
My doubt would be is it possible for the same create course to be fired the second time while the first is still creating sessions, which would result in 2x or 3x (depends on how many times the course is copied) and an unjustifiable backend load?

ypinhsuan replied:
Copying feedback sessions is a part of `CopyingCourse`. When feedback sessions are still being copied, the copy button will be disabled so instructors cannot copy another course at the same time. But they can still add a new course with just course name and course id.

moziliar reviewed (commented):
Left some comments.

Generally it feels good to me!

ccyccyccy reviewed (commented):
Found a bug. Copying to a destination course id with the same course id as a currently existing course will lead to the page showing a success message even though it failed.
![ezgif-4-59d45c04fe03](https://user-images.githubusercontent.com/32454748/121769829-67b03200-cb98-11eb-96fa-0bb67500ea97.gif)

Also, the select all checkbox merges into the rest of it's selection. Can we separate them more? Something like this would be nice. 
![image](https://user-images.githubusercontent.com/32454748/121769927-ead18800-cb98-11eb-9da0-2b9397bdd71c.png)

ypinhsuan commented on src/web/app/components/copy-course-modal/copy-course-modal.component.scss:
```diff
@@ -0,0 +1,4 @@
+.modal-selection * {
```
Not used, I will remove this

ypinhsuan commented on src/web/app/components/copy-course-modal/copy-course-modal.component.ts:
```diff
@@ -0,0 +1,111 @@
+
+  ngOnInit(): void {
+    for (const [id, offset] of Object.entries(this.timezoneService.getTzOffsets())) {
+      const hourOffset: number = Math.floor(Math.abs(offset) / 60);
```
I will keep the type since lint is checking for typedef and I will fail the check if I remove them

ypinhsuan commented:
Thank you for all the comments! I have updated the UI accordingly and fixed the bugs.
![updates](https://user-images.githubusercontent.com/58677983/121855972-893f2400-cd26-11eb-852d-8914ced1d65e.gif)

ccyccyccy commented:
> ![updates](https://user-images.githubusercontent.com/58677983/121855972-893f2400-cd26-11eb-852d-8914ced1d65e.gif)

Can we simply disable the "Copy" button with a warning text beside the input box to avoid this situation all together?

ypinhsuan commented:
The input box will turn red if the course ID is already been used (includes archived and soft-deleted courses) and a toast message will be displayed. Users can only copy from active courses for now.
![disable copy button](https://user-images.githubusercontent.com/58677983/122341098-addd0b00-cf75-11eb-8446-fb406bbb4ba7.gif)

ccyccyccy commented on src/web/app/components/copy-course-modal/copy-course-modal.component.ts:
```diff
@@ -0,0 +1,124 @@
+      return;
+    }
+
+    this.hasCourseId = this.allCourses.filter((course: Course) => course.courseId === this.newCourseId).length > 0;
```
Can change variable name to `newCourseIdIsConflicting`. Currently is a little confusing.

ccyccyccy replied:
Also, this variable should be set back to false on focus on the input element.

ccyccyccy commented on src/web/app/components/copy-course-modal/copy-course-modal.component.scss:
```diff
@@ -0,0 +1,7 @@
+hr.solid-divider {
+  border-top: 1px solid #BBB;
```
Can override the margin here? The default margin is way too much. Something like 0.3rem?

ypinhsuan replied:
I have reduced the margin for the border to look like this:
![ezgif com-gif-maker](https://user-images.githubusercontent.com/58677983/122697317-a8363c80-d277-11eb-8f4d-a4dd60163a19.gif)

ccyccyccy commented on src/web/app/components/copy-course-modal/copy-course-modal.component.ts:
```diff
@@ -0,0 +1,124 @@
+
+  ngOnInit(): void {
+    Object.entries(this.timezoneService.getTzOffsets())
+      .map(([id, offset]: [string, number]) => {
```
When I mentioned using `.map`, i meant something like `this.timezones = Object.entries(this.timezoneService.getTzOffsets()).map(...)`, so that you don't have to use `.push()`

ccyccyccy reviewed (commented):
Just some minor issues for me. Otherwise, all looks good.

rrtheonlyone commented:
For your kind review: either @ccyccyccy or @moziliar, when you have time. Let me have a final review before merging.

rrtheonlyone reviewed (changes requested):
Nice one 💯 I like the UI here and the frontend looks good.

There are two aspects that I am not a fan of though:
1. We are being too fine-grained in the copying. We should not need to make this many round-trips. 
2. We are duplicating the feedback session copying logic in 2 places. This is prone to errors down the road.

**Important**: We also need to update E2E tests in `InstructorCoursesPageE2ETest` for this feature before we can merge in.

This is my proposal for this. Let me know what you think @ypinhsuan 

## Optimized Copying

### Method

1. We should augment the `CreateFeedbackSessionAction` endpoint to have an additional field in the body e.g. like 'toCopySessionId` or something
2. Upon seeing this field the backend can then do this:
   - Get the old session and create a new session
   - Bulk get all questions corresponding to old session and bulk create new questions in new session
3. Return `FeedbackSessionData` 
4. We can then update `feedbackSessionsService` and use the new method in both places (no more duplication)

### Rationale

Step (2) should not be an expensive operation since there are only 4 database operations. Even if there are 1000 questions, I suspect it should easily finish within a reasonable amount of time (< 1 minute). However, this is only a guess and I do not have tests to prove it. It is worth a try though.

ypinhsuan commented:
I agree that there seem to be too many round-trips when copying a session. The optimized copying looks good and I will try implementing it.

ypinhsuan commented:
I have changed the way feedback sessions are copied and added e2e test for copying course.

rrtheonlyone commented on src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java:
```diff
@@ -1,19 +1,41 @@
+    }
+
     @Override
     public void validate() {
```
We need to update the validation - I think if `toCopyCourseId` is not null then `oldSessionName` should not be null as well?

Can we also rename `oldSessionName` to `toCopySessionName`?

ypinhsuan replied:
I have updated the validation and renamed to `toCopySessionName`

rrtheonlyone commented on src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java:
```diff
@@ -74,4 +84,47 @@ JsonResult execute() {
+                    .withQuestionDescription(question.getQuestionDescription())
+                    .build();
+
+            // validate questions (giver & recipient)
```
let's think about this for a second - why do we need to validate again if the old questions have already been validated? Tell me a scenario in which we run into a validation error here

ypinhsuan replied:
I removed the part that checks for the validity of giver and recipient.
If a feedback session is copied to a course with different number of students / teams, the `numOfGeneratedMsqChoices` would be different. So the part that validates question details is needed to update `numOfGeneratedMsqChoices` .

wkurniawan07 replied:
@damithc I need to raise this here. This is a problem I've been aware of for quite some time. (hello #8453)

This part here validates that the max or min number of choices (whichever ones configured) is/are within the range of the number of generated choices (student, team, instructor, etc.). At the first glance, it seems sensible. However, this creates an unnecessary coupling between the question and the course structure.

Consider this:
- A question with min. 5 choice, generated from instructors, is created, in a course with 6 instructors
- 2 instructors are removed from that course
- This question is now unanswerable, as the validation is done only during question creation/update, but not when the coupled course structure is changed

And in this very example (copying course), any MSQ question with minimum number of selected choice generated from students/teams is not copiable, because a new course doesn't have any student list to be generated from!

I suggest to just remove this validation.

damithc replied:
> I suggest to just remove this validation. Not because this validation creates the problem or removing this validation solves the problem, but because this validation doesn't solve the problem it intends to.

I agree. Let's do that.

rrtheonlyone reviewed (changes requested):
Excellent Work! Give me a ping (on slack) once you fix the request changes and I will prioritise the review immediately.

Minor comments below plus:
- Can we update `CreateFeedbackSessionActionTest` for the new behaviour we just introduced?
- Fix merge conflicts in the E2E tests

wkurniawan07 commented on src/web/app/components/copy-course-modal/copy-course-modal.component.html:
```diff
@@ -0,0 +1,77 @@
+<div class="modal-header">
```
Use 2 spaces for HTML file

wkurniawan07 commented on src/web/app/components/copy-course-modal/copy-course-modal.component.html:
```diff
@@ -0,0 +1,77 @@
+                <label>Course Name:</label>
+                <input id="copy-course-name" class="form-control" type="text" placeholder="e.g. Software Engineering" [(ngModel)]="newCourseName"
+                       [maxlength]="64"/>
+                <span>{{ 64 - newCourseName.length }} characters left</span>
```
`64` is a magic number

wkurniawan07 commented on src/main/java/teammates/ui/request/FeedbackSessionCreateRequest.java:
```diff
@@ -1,24 +1,47 @@
 
         assertTrue(feedbackSessionName != null, "Session name cannot be null");
         assertTrue(!feedbackSessionName.isEmpty(), "Session name cannot be empty");
+        assertTrue(toCopyCourseId == null || toCopySessionName != null, "To copy session name cannot be null");
```
- The validation message doesn't match
- "To copy session name" is not a noun (it's a verb). Should be "Session name to be copied from".

ypinhsuan replied:
Updated the message

wkurniawan07 commented on src/main/java/teammates/ui/webapi/CreateFeedbackSessionAction.java:
```diff
@@ -74,4 +82,37 @@ public JsonResult execute() {
+            try {
+                attributes = logic.createFeedbackQuestion(attributes);
+            } catch (InvalidParametersException e) {
+                throw new InvalidHttpRequestBodyException(e.getMessage(), e);
```
A valid question in one session should be valid in any other session, so if this happens, something is not right and it's not due to user. Need a `log.severe` here instead of throwing this exception.

wkurniawan07 commented on src/web/app/components/copy-course-modal/copy-course-modal.component.ts:
```diff
@@ -0,0 +1,122 @@
+      .filter((course: Course) => course.courseId === this.newCourseId).length > 0;
+    if (this.newCourseIdIsConflicting) {
+      this.statusMessageService.showErrorToast(
+        `The course ID ${this.newCourseId} has been used by another course, possibly by some other user.`);
```
This message makes it sound like you're comparing the name with all courses in the DB, not just all courses from the user. Is it really what's being done here?

ypinhsuan replied:
This part checks only for courses that belong to the instructor, I will update the message.

wkurniawan07 commented on src/web/app/pages-instructor/instructor-courses-page/instructor-courses-page.component.ts:
```diff
@@ -246,6 +274,130 @@ export class InstructorCoursesPageComponent implements OnInit {
+    }
+    const isLoadingCourseStats: boolean = false;
+    const courseModel: CourseModel = { course, canModifyCourse, canModifyStudent, isLoadingCourseStats };
+    return courseModel;
```
`courseModel` needs not be a variable? Can just return the previous line directly.

wkurniawan07 reviewed (changes requested):
Not really looking into detail, but I do have some questions/concerns.

rrtheonlyone reviewed (approved):
LGTM 💯 Thanks for the work put in here!
